### PR TITLE
Training: gradient clipping (max_grad_norm) + gradient accumulation

### PR DESCRIPTION
**Pull request by fxd0h**
OptimizerSpec gains `max_grad_norm`. When set, gradients are clipped by global norm (`mlx.optimizers.clip_grad_norm`) before `optimizer.update`. `None` = off, so existing configs are unchanged. Guards against gradient spikes / NaN blowups during LoRA training. Full fast suite: 516 passed.

Also adds `gradient_accumulation_steps` (default 1 = off). When > 1, grads are averaged over that many micro-batches and the optimizer steps once per window, for an effective batch of `batch_size * gradient_accumulation_steps` without the memory of a bigger batch. The step count and save/plot/preview frequencies stay in micro-batch units. Verified on z-image: 5 micro-batches give 2 optimizer steps at N=2 and 5 at N=1 (no regression).

## Summary by CodeRabbit

* **New Features**
  * Added configurable gradient accumulation for more flexible training batch sizes.
  * Added optional global gradient clipping to improve training stability.

* **Bug Fixes**
  * Improved training reliability by preserving partial accumulations appropriately and safely resetting them after non-finite losses.

<details open><summary><h3>Greptile Summary</h3></summary>

The PR adds configurable global-norm gradient clipping and gradient accumulation while preserving existing behavior through disabled defaults.
- Adds `max_grad_norm` and `gradient_accumulation_steps` optimizer configuration.
- Averages gradients across valid micro-batches and resets the complete partial window after nonfinite losses.
- Adds focused tests for clean, skipped, and incomplete accumulation windows.
</details>

<h3>Confidence Score: 5/5</h3>

The PR appears safe to merge.

No blocking failure remains.

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/mflux/models/common/training/state/training_spec.py | Adds backward-compatible optimizer settings for optional gradient clipping and accumulation. |
| src/mflux/models/common/training/trainer.py | Implements valid-micro-batch accumulation, synchronized reset after nonfinite losses, optional clipping, and materialization of partial sums. |
| tests/training/test_training_trainer.py | Covers clean accumulation windows, reset-and-extension after skipped batches, and intentional dropping of incomplete windows. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Process micro-batch] --> B{Finite loss?}
    B -- No --> C[Reset accumulated gradients and valid count]
    B -- Yes --> D[Scale and fold gradients into window]
    D --> E{Window complete?}
    E -- No --> F[Materialize partial gradient sum]
    E -- Yes --> G{Clipping enabled?}
    G -- Yes --> H[Clip by global norm]
    G -- No --> I[Optimizer update]
    H --> I
    I --> J[Evaluate parameters and optimizer state]
```
</details>

<sub>Reviews (4): Last reviewed commit: ["Merge branch &#39;main&#39; into pr/training-gra..."](https://github.com/mflux-community/mflux/commit/06ce8ba2d17f5e16a4bdfd92aa7c650356055f46) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=51896714)</sub>

### CODE REVIEW

File: src/mflux/models/common/training/trainer.py [hunk 1]
```diff
@@ -161,18 +162,44 @@ def train(
...
+            # Gradient accumulation: average grads across accum_steps micro-batches and only step
+            # the optimizer on the window boundary, for an effective batch of batch_size *
+            # accum_steps. num_iterations counts micro-batches, so the boundary is every
+            # accum_steps of them; bookkeeping below still runs each iteration on valid weights.
+            at_step_boundary = training_state.iterator.num_iterations % accum_steps == 0
+            if accum_steps > 1:
+                grads = tree_map(lambda g: g / accum_steps, grads)
+                if accumulated_grads is not None:
+                    grads = tree_map(lambda a, g: a + g, accumulated_grads, grads)
+                accumulated_grads = None if at_step_boundary else grads
```
[the remark this answers is not in the document]
Comment (fxd0h) [reply src/mflux/models/common/training/trainer.py:194]: Both of you found the same thing and it was real: with `accum_steps` 4 and micro-batch 2 non-finite, the optimizer stepped on two gradients still divided by four, half the update either setting asks for, and nothing said so.

Fixed in 4450b4e. The window now counts valid micro-batches and resets that count together with the accumulated sum, so a skip extends the window instead of closing it early. The bookkeeping moved into a small helper so it could be tested directly: three cases cover a clean window, a window a skip extends, and one that never completes.

File: src/mflux/models/common/training/trainer.py [hunk 2]
```diff
@@ -161,18 +162,44 @@ def train(
...
-            training_state.optimizer.optimizer.update(model=adapter.model(), gradients=grads)
-            mx.eval(adapter.model().parameters(), training_state.optimizer.optimizer.state)
-            del loss, grads
+            del loss
+
+            # Gradient accumulation: average grads across accum_steps micro-batches and only step
+            # the optimizer on the window boundary, for an effective batch of batch_size *
+            # accum_steps. num_iterations counts micro-batches, so the boundary is every
+            # accum_steps of them; bookkeeping below still runs each iteration on valid weights.
+            at_step_boundary = training_state.iterator.num_iterations % accum_steps == 0
```
[the remark this answers is not in the document]
Comment (fxd0h) [reply src/mflux/models/common/training/trainer.py]: Both of you found the same thing and it was real: with `accum_steps` 4 and micro-batch 2 non-finite, the optimizer stepped on two gradients still divided by four, half the update either setting asks for, and nothing said so.

Fixed in 4450b4e. The window now counts valid micro-batches and resets that count together with the accumulated sum, so a skip extends the window instead of closing it early. The bookkeeping moved into a small helper so it could be tested directly: three cases cover a clean window, a window a skip extends, and one that never completes.

File: src/mflux/models/common/training/trainer.py [hunk 3]
```diff
@@ -161,18 +162,44 @@ def train(
...
+
+            if accum_steps == 1 or at_step_boundary:
+                if max_grad_norm is not None:
+                    grads, _ = clip_grad_norm(grads, max_grad_norm)
+                training_state.optimizer.optimizer.update(model=adapter.model(), gradients=grads)
+                mx.eval(adapter.model().parameters(), training_state.optimizer.optimizer.state)
+            else:
+                # Keep the partial sum materialized so the graph doesn't grow across the window.
+                mx.eval(accumulated_grads)
+            del grads
```
[the remark this answers is not in the document]
Comment (fxd0h) [reply src/mflux/models/common/training/trainer.py:204]: Not taking this one. Dropping the final partial window is the same choice as `drop_last` in a dataloader: an update built from fewer micro-batches than the window asks for is a different-sized step, and taking it at the end of a run buys one slightly-off update in exchange for making the effective batch size not quite what the config says.

The resume half is real but it belongs to the resume machinery rather than here. Persisting a partial window means writing gradients into the checkpoint, which is new state to version and restore, for the benefit of at most `accum_steps - 1` micro-batches per interruption. Worth its own issue if someone wants it; it would be a larger change than this PR.

**Comment by fxd0h** [comment]
Left as is. The codebase documents non-obvious helpers this way, and what the docstring carries here is the reason the window counts valid micro-batches rather than iterations, which is the part a reader cannot get from the name.

File: src/mflux/models/common/training/trainer.py [hunk 4]
```diff
@@ -161,18 +162,44 @@ def train(
...
-            del loss, grads
+            del loss
+
+            # Gradient accumulation: average grads across accum_steps micro-batches and only step
+            # the optimizer on the window boundary, for an effective batch of batch_size *
+            # accum_steps.
+            at_step_boundary = True
+            if accum_steps > 1:
+                grads, accumulated_count, at_step_boundary = TrainingTrainer._fold_into_window(
+                    grads, accumulated_grads, accum_steps, accumulated_count
```
[the remark this answers is not in the document]
Comment (fxd0h) [reply src/mflux/models/common/training/trainer.py:192]: This is the same call as the CodeRabbit thread above, and I am making it the same way. Dropping an incomplete window is what `drop_last` does in a dataloader: an update built from fewer micro-batches than the window asks for is a differently sized step, and taking it at the end of a run buys one slightly-off update at the cost of making the effective batch size something other than what the config says.

The part worth keeping is the resume case, where a run stopped mid-window loses the micro-batches it had already spent. That belongs to the checkpoint format rather than to this loop, since fixing it means persisting the partial sum and the count alongside the optimizer state.

File: src/mflux/models/common/training/trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fxd0h) [reply src/mflux/models/common/training/trainer.py:204]: No follow-up issue for now. Persisting a partial window is only worth doing together with the rest of the resume state, and that is a change to the checkpoint format rather than something to track on its own.

**Review by ianscrivener — APPROVED** [review]
<sub>Approved in [CodeRabbit Change Stack](https://app.coderabbit.ai/change-stack/mflux-community/mflux/pull/447)</sub>

**Comment by ianscrivener** [comment]
@copilot resolve the merge conflicts in this pull request
